src/test/java/io/confluent/examples/streams/CustomStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,538 @@
+ * manage the table-side store manually (because you use a KStream instead of a KTable for the table's data).
+ */
+public class CustomStreamTableJoinIntegrationTest {
+
+  @ClassRule
+  public static final EmbeddedSingleNodeKafkaCluster CLUSTER = new EmbeddedSingleNodeKafkaCluster();
+
+  private static final String inputTopicForStream = "inputTopicForStream";
+  private static final String inputTopicForTable = "inputTopicForTable";
+  private static final String outputTopic = "outputTopic";
+  private static final String tableStoreName = "table-store";
+
+  @BeforeClass
+  public static void startKafkaCluster() {
+    CLUSTER.createTopic(inputTopicForStream);
+    CLUSTER.createTopic(inputTopicForTable);
+    CLUSTER.createTopic(outputTopic);
+  }
+
+  @Test
+  public void shouldTriggerStreamTableJoinFromTable() throws Exception {
+    final Duration approxMaxWaitTimePerRecordForTableData = Duration.ofSeconds(5);
+    final Duration frequencyToCheckForExpiredWaitTimes = Duration.ofSeconds(2);
+
+    final List<KeyValueWithTimestamp<String, Double>> inputStreamRecords = Arrays.asList(
+      new KeyValueWithTimestamp<>("alice", 999.99, TimeUnit.MILLISECONDS.toMillis(10)),
+      new KeyValueWithTimestamp<>("bobby", 222.22, TimeUnit.MILLISECONDS.toMillis(15)),
+      new KeyValueWithTimestamp<>("alice", 555.55, TimeUnit.MILLISECONDS.toMillis(30)),
+      new KeyValueWithTimestamp<>("alice", 666.66, TimeUnit.MILLISECONDS.toMillis(40))
+    );
+
+    final List<KeyValueWithTimestamp<String, Long>> inputTableRecords = Arrays.asList(
+      new KeyValueWithTimestamp<>("alice", 1L, TimeUnit.MILLISECONDS.toMillis(20)),
+      new KeyValueWithTimestamp<>("alice", 2L, TimeUnit.MILLISECONDS.toMillis(39)),
+      new KeyValueWithTimestamp<>("bobby", 8L,
+        approxMaxWaitTimePerRecordForTableData.plus(Duration.ofSeconds(1)).toMillis())
+    );
+
+    final List<KeyValue<String, Pair<Double, Long>>> expectedOutputRecords = Arrays.asList(
+      new KeyValue<>("alice", new Pair<>(999.99, 1L)),
+      new KeyValue<>("alice", new Pair<>(555.55, 1L)),
+      new KeyValue<>("alice", new Pair<>(666.66, 2L)),
+      new KeyValue<>("bobby", new Pair<>(222.22, null))
+    );
+
+    //
+    // Step 1: Define and start the processor topology.
+    //
+    final StreamsBuilder builder = new StreamsBuilder();
+
+    final Properties streamsConfiguration = new Properties();
+    streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "custom-join-integration-test");
+    streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+    streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+    // Use a temporary directory for storing state, which will be automatically removed after the test.
+    streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+
```

Comment:
@mjsax: You asked about my experiments with non-zero settings for `max.task.idle.ms`. I unsuccessfully tried various settings (10, 100, 1000, 10000): this integration test always failed when I removed `withCachingDisabled` from the table's internal state store (see my comment at https://github.com/confluentinc/kafka-streams-examples/pull/232/files#r272945515), regardless of what `max.task.idle.ms` was set to.

```java
streamsConfiguration.put(StreamsConfig.MAX_TASK_IDLE_MS_CONFIG, 1000);
```

To reproduce:

1. Change

```java
Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as(tableStoreName).withCachingDisabled());
```
to
```
Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as(tableStoreName));
```

2. Try to achieve the same effect with `max.task.idle.ms.config`, e.g. by adding:

```
streamsConfiguration.put(StreamsConfig.MAX_TASK_IDLE_MS_CONFIG, 1000);
```

3. Run this integration test.